src/Cli/dotnet/ReleasePropertyProjectLocator.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Parsing;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Build.Execution;
+using Microsoft.DotNet.Cli.Sln.Internal;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools;
+using Microsoft.DotNet.Tools.Common;
+
+namespace Microsoft.DotNet.Cli
+{
+    class ReleasePropertyProjectLocator : ProjectLocator
+    {
+        /// <param name="slnProjectPropertytoCheck">A property to enforce if we are looking into SLN files. If projects disagree on the property, throws exception.</param>
+        /// <returns>A project instance that will be targeted to publish/pack, etc. null if one does not exist.</returns>
+        public override ProjectInstance GetTargetedProject(IEnumerable<string> slnOrProjectArgs, string slnProjectPropertytoCheck = "")
+        {
+            string potentialProject = "";
+
+            foreach (string arg in slnOrProjectArgs.Append(Directory.GetCurrentDirectory()))
+            {
+                if (IsValidProjectFilePath(arg))
+                {
+                    return TryGetProjectInstance(arg);
+                }
+                else if (Directory.Exists(arg)) // We should get here if the user did not provide a .proj or a .sln
+                {
+                    try
+                    {
+                        return TryGetProjectInstance(MsbuildProject.GetProjectFileFromDirectory(arg).FullName);
+                    }
+                    catch (GracefulException)
+                    {
+                        string potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
+
+                        if (!string.IsNullOrEmpty(potentialSln))
+                            return GetSlnProject(potentialSln, slnProjectPropertytoCheck);
+                    } // If nothing can be found: that's caught by MSBuild XMake::ProcessProjectSwitch -- don't change the behavior by failing here. 
+                }
+            }
+
+            return string.IsNullOrEmpty(potentialProject) ? null : TryGetProjectInstance(potentialProject);
+        }
+
+        /// <returns>The top-level project (first if multiple exist) in a SLN. Returns null if no top level project. Throws exception if two top level projects disagree
+        /// in the configuration property to check.</returns>
+        public override ProjectInstance GetSlnProject(string slnPath, string slnProjectConfigPropertytoCheck = "")
+        {
+            if (Environment.GetEnvironmentVariable("ENABLE_P_RELEASE_SLN") == null) // This has a performance overhead so don't do this unless opted in.
```

Comment:
This can also take effect for pack and not just publish, what is the best naming convention for that? Or do we want two separate ones? 
`DOTNET_CLI_ENABLE_PUBLISH_RELEASE_FOR_SOLUTIONS` 
`DOTNET_CLI_ENABLE_PACK_OR_PUBLISH_RELEASE_FOR_SOLUTIONS`